### ISSUE TITLE
optimize @evalpoly for SIMD?

### ISSUE THREAD

**Issue by stevengj** [JuliaLang/julia#18410]
Right now, `@evalpoly(x, coefs...)` uses Horner's rule for real `x`, and a more complicated algorithm for complex `x`, and is performance-critical for evaluation of special functions.

However, it seems likely to be faster to try and exploit SIMD instructions for these polynomial evaluations, e.g. with [Estrin's algorithm](https://en.wikipedia.org/wiki/Estrin%27s_scheme) or some other technique.    (The nice thing about code generation / macros is that we can investigate fancy methods for this sort of thing.)

**Comment by yuyichao** [comment]
Will the new algorithm still be optimum when used in a vectorized loop?

**Comment by ViralBShah** [comment]
This may be relevant for the new libm.jl work. Cc @musm

**Comment by stevengj** [comment]
@yuyichao, probably?  Even if the compiler inlines a special-function call like `erfinv`, I'm skeptical that compilers are powerful enough to vectorize the polynomial evaluation.  e.g. I don't see any evidence of this in:
```jl
julia> function foo(a)
          s = zero(eltype(a))
          @simd for i in eachindex(a)
              s += @inbounds @evalpoly(a[i], 2.3, 4.2, 1.6, 8.9, 0.8)
          end
          return s
       end
foo (generic function with 1 method)

julia> @code_native foo(rand(10))
```
if I'm reading the asm output correctly.

**Comment by yuyichao** [comment]
I'm not worrying about inlining special functions that use `@evalpoly`, I'm worrying about the case where `@evalpoly` is used directly in the loop.

The code you posted has two problems.

1. It's accumulating `nothing`.

    Until https://github.com/JuliaLang/julia/pull/15558 is merged you need to put the `@inbounds` on the loop or before the assignment.

2. It does seem that the loop body is somehow too complicated for llvm to vectorize in this case (might worth investigating) However, if you are doing purely elementwise operations, a very valid use case IMHO and is even more relavant now with the broadcast syntax, i.e.

```jl
julia> function foo2(a)
           s = zero(eltype(a))
           @simd for i in eachindex(a)
               @inbounds a[i] = @evalpoly(a[i], 2.3, 4.2, 1.6, 8.9, 0.8)
           end
           return s
       end
```

Then LLVM vectorizes it without any problem (3.8.1 at least)

P.S. I strongly recommend checking `code_llvm` even for people who know assembly well. It's much easier to notice the problem with `@inbounds` in the llvm ir.

**Comment by stevengj** [comment]
Ah, thanks.   Even so, it might be nice to have a SIMD variant of `@evalpoly` for use in special functions.

**Comment by yuyichao** [comment]
> for use in special functions.

Or cases not meant to/possible to be vectorized in a loop in general, I agree. It would be nice if the new algorithm won't regret in the vectorize loop case but I guess the answer is that it might?

**Comment by musm** [comment]
I got bored in class

only works for odd polynomials. I don't think simd is being attached however? could anyone confirm. 
```julia
#only for odd polynomials
#c in descending order
function estrin(x,c)
n = length(c)
m = n÷2
X = x
ap = c
while n > 1
    global a = Array(Float64,m)
    @simd for i = 1:m
        @inbounds a[m+1-i] = muladd(ap[n+1-(2i-1)], X, ap[n+1-(2i-1)-1])
    end
    X = X*X
    n = m
    m = n÷2
    ap = a
end
return @inbounds a[1]
end
@show estrin(x,c)
```

**Comment by yuyichao** [comment]
What are you trying to achieve with globals?

**Comment by musm** [comment]
My bad, fixed, thanks!

**Comment by yuyichao** [comment]
There's also unnecessary copy/allocation. `ap` should be updated in place.

**Comment by yuyichao** [comment]
Or even better, IIUC, you can just swap the two buffers.

**Comment by yuyichao** [comment]
I don't think you can get vectorization with non-unit stride (there's trick to kind of get it working though).

I just realize this is the evalpoly issue (I thought this is one of the special/libm function ones......) in this case any allocation is already a huge lost. `@evalpoly` generates an expression at parse time and the replacement should do that too.

**Comment by musm** [comment]
would 
```julia
    @inbounds ap[1:m] = a[1:m]
```
swap the buffers? I might be misunderstanding what you are saying

**Comment by yuyichao** [comment]
`ap, a = a, ap`

`ap[1:m] = a[1:m]` still allocating a new array and copying the data. `ap[:] = a` won't allocate, but will still copy.

**Comment by stevengj** [comment]
@musm, rather than being "far more efficient", in a quick benchmark I find your `estrin` function to be about 50x slower than the inlined polynomial evaluation of `@evalpoly` for this degree-7 polynomial.   This is not surprising, because the difference between inlined polynomial evaluation and any kind of runtime loop is vast.

This issue is about improving the inlined `@evalpoly`; let's not side-track it into a discussion of how to optimize some runtime-loop code.

**Comment by musm** [comment]
apologies

**Comment by stevengj** [comment]
No worries @musm, I just wanted to nip the digression in the bud.

**Comment by musm** [comment]
On the topic of improving `@evalpoly`

> However, it seems likely to be faster to try and exploit SIMD instructions for these polynomial evaluations, e.g. with Estrin's algorithm or some other technique.

I'm not very strong with macros but I now have a pretty decent implementation of Estrin's algorithm that's a regular function  (for odd polynomials, still). I have  two  versions: one without simd which is 10x slower than @horner and one that takes advantage of simd which is  40 times slower, based on benchmarks.

It's not  exactly clear how one could turn this algorithm   to  a macro that would expand to avoid runtime loops and take advantage of simd operations. If I understand things, would it be possible to expand the macro to something like the following

```julia

@simd begin
a1 = c2*x + c1 
a2 = c4*x + c3
a3 = c6*x + c5
a4 = c8*x + c7
end
x = x*x
@simd begin
a1 = a2*x + a1 
a2 = a4*x + a3
end
@simd begin
a1 = a2*x + a1 
end
return a1
```

**Comment by yuyichao** [comment]
`@simd` used this way is a no-op. The vectorization here requires writing it in a way that can be recognized by LLVM passes we add with `-O3` or explicitly using `VecElement`.

**Comment by eschnett** [comment]
The [SIMD](https://github.com/eschnett/SIMD.jl) package provides the respective operations, based on `Base.VecElement`:
```Julia
using SIMD
function evalpoly{T}(c1,c2,c3,c4,c5,c6,c7,c8, x::T)
    c2468 = Vec{4,T}((c2,c4,c6,c8))
    c1357 = Vec{4,T}((c1,c3,c5,c7))
    a1234 = c2468 * x + c1357
    x = x^2
    a24 = Vec{2,T}((a1234[2], a1234[4]))
    a13 = Vec{2,T}((a1234[1], a1234[3]))
    a12 = a24 * x + a13
    x = x^2
    a2 = Vec{1,T}((a12[2],))
    a1 = Vec{1,T}((a12[1],))
    r = a2 * x + a1
    r
end

evalpoly(1.0,2.0,3.0,4.0,5.0,6.0,7.0,8.0, 0.5)
@code_native evalpoly(1.0,2.0,3.0,4.0,5.0,6.0,7.0,8.0, 0.5)
```

If you look at the generated code, you'll see immediately why this turns out to be very slow:
```
julia> @code_native evalpoly(1.0,2.0,3.0,4.0,5.0,6.0,7.0,8.0, 0.5)
       	.section       	__TEXT,__text,regular,pure_instructions
Filename: REPL[20]
       	pushq  	%rbp
       	movq   	%rsp, %rbp
       	vmovsd 	16(%rbp), %xmm8         ## xmm8 = mem[0],zero
Source line: 121
       	vunpcklpd      	%xmm7, %xmm5, %xmm5 ## xmm5 = xmm5[0],xmm7[0]
       	vunpcklpd      	%xmm3, %xmm1, %xmm1 ## xmm1 = xmm1[0],xmm3[0]
       	vinsertf128    	$1, %xmm5, %ymm1, %ymm1
Source line: 121
       	vunpcklpd      	%xmm6, %xmm4, %xmm3 ## xmm3 = xmm4[0],xmm6[0]
       	vunpcklpd      	%xmm2, %xmm0, %xmm0 ## xmm0 = xmm0[0],xmm2[0]
       	vinsertf128    	$1, %xmm3, %ymm0, %ymm0
Source line: 114
       	vbroadcastsd   	%xmm8, %ymm2
Source line: 567
       	vmulpd 	%ymm2, %ymm1, %ymm1
Source line: 567
       	vaddpd 	%ymm1, %ymm0, %ymm0
Source line: 5
       	vmulsd 	%xmm8, %xmm8, %xmm1
Source line: 121
       	vextractf128   	$1, %ymm0, %xmm2
Source line: 121
       	vunpckhpd      	%xmm2, %xmm0, %xmm3 ## xmm3 = xmm0[1],xmm2[1]
Source line: 121
       	vunpcklpd      	%xmm2, %xmm0, %xmm0 ## xmm0 = xmm0[0],xmm2[0]
Source line: 114
       	vmovddup       	%xmm1, %xmm2    ## xmm2 = xmm1[0,0]
Source line: 567
       	vmulpd 	%xmm3, %xmm2, %xmm2
Source line: 567
       	vaddpd 	%xmm2, %xmm0, %xmm0
Source line: 9
       	vmulsd 	%xmm1, %xmm1, %xmm1
Source line: 121
       	vpermilpd      	$1, %xmm0, %xmm2 ## xmm2 = xmm0[1,0]
Source line: 567
       	vmulsd 	%xmm2, %xmm1, %xmm1
Source line: 567
       	vaddsd 	%xmm1, %xmm0, %xmm0
Source line: 13
       	popq   	%rbp
       	vzeroupper
       	retq
```

There are fewer actual arithmetic operations (`vmul`, `vadd`), but there are many additional operations to move data between vector lanes.

The first six operations (`vunpcklpd` and `vinsertf128`) are merely setting up the constants `c1`...`c8`; these would normally be known at compile time, making those instructions superfluous. The other `vextract`, `vunpck`, `vmov`, and `vperm` instructions remain necessary.

**Comment by musm** [comment]
Thanks @eschnett  I created a macro that expands out the operation's in Estrin's algorithm to explore  possible benefits of vectorization for polynomial evaluation using your SIMD package. See:
https://github.com/musm/Estrin.jl  (I might rename it later)
Do you mind taking a look? The code works (`macroexpand` looks fine), but something's terribly wrong as can be seen by the  `@code_native` . Sorry for stealing the discussion again, comments on the code should be made in that repo.

**Comment by eschnett** [comment]
@musm Nice!

The expensive part is currently switching between different vector sizes. The following might lead to a faster algorithm:
- pick a particular vector size, e.g. 4, and use it for all operations
- if you need a larger vector size, express this via loops over 4-vectors, maybe using manually unrolled loops
- if you need a smaller vector size, then pad the operations, e.g. duplicating operations
- for the last (scalar) operation, use a scalar
- keep `x` also in a vector register, so that it isn't cast from scalar to vector multiple times; this means squaring it as vector
The goal would be to get things down to a single vector permute operation in between the different iterations.

However, if this `evalpoly` is called inside a loop, then I expect things to be much faster if `evalpoly` is kept scalar, and the loop is vectorized.

**Comment by musm** [comment]
@eschnett I'll try to take those points into consideration in an improved implementation.

Another very simple option for parallelism is to do an even odd split of the polynomial.



```julia

using BenchmarkTools
using Base.Math.@horner


c0, c1, c2, c3, c4, c5, c6, c7, c8, c9, c10 = (1:11...)
x = 2.0


function f1(x, c0, c1, c2, c3, c4, c5, c6, c7, c8, c9, c10)
x2 = x*x
begin # ideally we should be able to tell the compiler to run these two in parallel, is threads the right approach here? I.e. run these two on separate threads
cc1 = @horner x2 c1 c3 c5 c7 c9
cc2 = @horner x2 c2 c4 c6 c8 c10
end
c0 + x*cc1 + x2 * cc2
end

function f2(x, c0, c1, c2, c3, c4, c5, c6, c7, c8, c9, c10)
@horner x c0 c1 c2 c3 c4 c5 c6 c7 c8 c9 c10
end

@show @benchmark f1(x, c0, c1, c2, c3, c4, c5, c6, c7, c8, c9, c10)
@show @benchmark f2(x, c0, c1, c2, c3, c4, c5, c6, c7, c8, c9, c10)

```

On my computer with -O3 this gives
```
julia> @show @benchmark f1(x, c0, c1, c2, c3, c4, c5, c6, c7, c8, c9, c10)

BenchmarkTools.Trial:
  samples:          10000
  evals/sample:     1000
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  16.00 bytes
  allocs estimate:  1
  minimum time:     42.00 ns (0.00% GC)
  median time:      44.00 ns (0.00% GC)
  mean time:        49.59 ns (2.86% GC)
  maximum time:     2.78 μs (97.30% GC)

julia> @show @benchmark f2(x, c0, c1, c2, c3, c4, c5, c6, c7, c8, c9, c10)

BenchmarkTools.Trial:
  samples:          10000
  evals/sample:     796
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  160.00 bytes
  allocs estimate:  10
  minimum time:     201.00 ns (0.00% GC)
  median time:      226.00 ns (0.00% GC)
  mean time:        243.06 ns (5.59% GC)
  maximum time:     3.94 μs (88.90% GC)
```

**Comment by ViralBShah** [comment]
Julia will not yet use multiple threads for concurrent evaluation.

**Comment by yuyichao** [comment]
It's very unlikely cheap operations like this will benefit from threading at all.

**Comment by ViralBShah** [comment]
@musm The time difference seems to be due partially to much higher gc. Can you either call `gc()` before each run, or do `gc_disable()` and run the tests?

**Comment by ViralBShah** [comment]
Also, I wonder if having all the variables in global scope affects this. Perhaps wrap everything in a function?

**Comment by yuyichao** [comment]
`g1.($xx)` instead of `g1.(xx)` should get rid of the global variable although I doubt it matters much.
And it's better to not use the boardcast form since as mentioned before it's the form that can be more easily loop vectorized with the old/scalar form. Using the broadcast from also means that you are sensitive to many unrelated factors including memory allocation, garbage collection and memory bandwidth.

**Comment by yuyichao** [comment]
> although I doubt it matters much.

I take that back, I didn't realize that all the `c*`s are in global scope. Yeah, that's terrible.....

**Comment by KristofferC** [comment]
If you see that you are allocating 300 MB in your benchmark that should just evaluate a polynomial then you should look closer in how you are benchmarking. The absolute timings you are getting are also unreasonable with some napkin math of cpu freq and number of ops required.

**Comment by stevengj** [comment]
Because the coefficients are known at parse time in the macro, it should be able to decide on the optimal vector size at parse time.   There should be no runtime switching.

**Comment by stevengj** [comment]
(@musm, this kind of low-level performance hacking is probably not a good introductory issue, because it requires a lot of understanding of Julia internals.)

**Comment by eschnett** [comment]
An even/odd splitting like this is likely to increase performance. Modern CPUs can execute instructions out-of-order, and given the number of execution ports they have, it is likely that both streams can be executed simultaneously.

Alternatively, you can execute the two streams simultaneously by using 2-element SIMD vectors. That's very similar to what I suggested earlier, except with a vector size of 2 instead of 4.

As others said, multi-threading will not help here. Threading introduces an overhead of about 1 us (roughly), whereas instructions are executed in fractions of a nanosecond. Thus you'll need tens of thousands of instructions to recoup the threading overhead.

**Comment by musm** [comment]
Sorry about the lack of rigorous benchmarking, I should've known better, but It was late : P 

@eschnett I don't see any strong evidence that it may be sufficient just to expand into and even and odd pair and let the CPU execute the instructions out of order.

To test this I implemented a macro that does an even odd split. You can do `Pkg.clone("https://github.com/musm/ParPoly.jl.git")`

warning: Not rigorous (since this only benchmarks a 9 degree polynomial)
```julia

using BenchmarkTools
using Base.Math.@horner
using ParPoly

const c0 = 1.0
const c1 = 2.0
const c2 = 3.0
const c3 = 4.0
const c4 = 5.0
const c5 = 6.0
const c6 = 7.0
const c7 = 8.0
const c8 = 9.0
const c9 = 10.0
const c10 = 11.0

x = 10*randn(1_000_000)

f1(x) = @horner_split x c0 c1 c2 c3 c4 c5 c6 c7 c8 c9
f2(x) = @horner x c0 c1 c2 c3 c4 c5 c6 c7 c8 c9

gc()
@benchmark f1.($x)
gc()
@benchmark f2.($x)

```

On my computer with -O3
```julia
julia> gc()

julia> @benchmark f1.($x)
BenchmarkTools.Trial:
  samples:          1083
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  7.63 mb
  allocs estimate:  3
  minimum time:     3.08 ms (0.00% GC)
  median time:      3.55 ms (0.00% GC)
  mean time:        4.61 ms (20.41% GC)
  maximum time:     18.04 ms (19.12% GC)

julia> gc()

julia> @benchmark f2.($x)
BenchmarkTools.Trial:
  samples:          1079
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  7.63 mb
  allocs estimate:  3
  minimum time:     3.14 ms (0.00% GC)
  median time:      3.52 ms (0.00% GC)
  mean time:        4.62 ms (20.39% GC)
  maximum time:     17.17 ms (57.97% GC)
```

> Alternatively, you can execute the two streams simultaneously by using 2-element SIMD vectors. That's very similar to what I suggested earlier, except with a vector size of 2 instead of 4.

If I understand correctly do you mean something like the following?
```julia

x1 = 2.0
x2 = x1*x1
x1x2 = Vec{2,Float64}((x1,x2))
xx   = Vec{2,Float64}((x2,x2))
cc1  = Vec{2,Float64}((c1,c2))
cc2  = Vec{2,Float64}((c3,c4))
cc3  = Vec{2,Float64}((c5,c6))
cc4  = Vec{2,Float64}((c7,c8))
cc5  = Vec{2,Float64}((c9,c10))
C = (cc1+xx*(cc2 + xx*(cc3 + xx*(cc4 + xx*cc5))))
val = c0 + sum(x1x2*C)
```
This would need to allocate a lot of new variables (not sure such a large number of allocation would outweigh the possible  increase of execution speed) I would think that a more parallelized version would benefit stronger from such an approach a la estrin's scheme or similar.

**Comment by yuyichao** [comment]
Again, don't use the broadcast to benchmark, use the scalar form directly. Also those `Vec`s don't allocate at all.

**Comment by musm** [comment]
Ah yes thank you. I updated the benchmarks to more rigorously test and compare the results for different polynomial orders. 

![image](https://cloud.githubusercontent.com/assets/4319522/18412654/3a4c90d6-7762-11e6-95f1-a5dcea156270.png)

**Comment by yuyichao** [comment]
`@benchmark` is good enough, don't write your own loop..................................................

**Comment by yuyichao** [comment]
I also believe that you should be seeing the function overwrite warning and you should no ignore it. You are most likely benchmarking only a single function which is why the scaling in your result doesn't make sense.

**Comment by musm** [comment]
@yuyichao Does that look better?

**Comment by yuyichao** [comment]
> what's the problem then

#265

**Comment by musm** [comment]
Wasn't aware, thank you....ok so after all the mistakes I have made I hope this final result is a good benchmark

![image](https://cloud.githubusercontent.com/assets/4319522/18412773/64ac7d8e-7765-11e6-8933-895c11660531.png)

**Comment by yuyichao** [comment]
The time the low order polynomials is way to long. You shouldn't get tens of ns for that unless you are using a really slow CPU. Try,

```jl
for n in nn
    c = randn(Float64,n)
    @gensym f1
    @gensym f2
    @eval begin
        $f1(x) = @horner_split x $(c...)
        $f2(x) = @horner x $(c...)
        t_hs = @benchmark $f1($x)
        t_h  = @benchmark $f2($x)
    end
    push!(t_horner_split, median(t_hs).time)
    push!(t_horner, median(t_h).time)
end
```

**Comment by StefanKarpinski** [comment]
@musm, good investigation – benchmarking this kind of stuff is tricky and this is how you learn :)

**Comment by yuyichao** [comment]
It does seem that you might hit https://github.com/JuliaCI/BenchmarkTools.jl/issues/7, which is why I still use `Benchmarks.jl` instead `BenchmarkTools.jl` for small functions like this.....

**Comment by musm** [comment]
I think so too actually. These are in the ns range. Anecdotally using Benchmarks.jl @horner_split edges out @horner especially for larger polynomial sizes. I can't however seem to  get `@benchmark` working in an `@eval`  I keep getting
```julia
ERROR: InexactError()
 in trunc(::Type{Int64}, ::Float64) at .\float.jl:463
 in execute(::###494, ::Int64, ::Int64, ::Float64, ::Float64, ::Int64, ::Bool) at C:\Users\Mustafa\.julia\v0.5\Benchmarks\src\execute.jl:62
 in execute(::Function) at C:\Users\Mustafa\.julia\v0.5\Benchmarks\src\execute.jl:42
 in eval(::Module, ::Any) at .\boot.jl:234
 in macro expansion; at .\REPL[52]:12 [inlined]
 in anonymous at .\<missing>:?
```

**Comment by musm** [comment]
I've been doing some more benchmarking and I see very good results with the simple @horner_split, in various test cases it's usually 2x faster than the @horner method in base. This can probably be attributed to the fact that the cpu will just execute the two even and odd lines out of order.  Interestingly, the @horner_simd method is according to  my tests shows up as slower   than the @horner_split method and the version in base.

**Comment by ViralBShah** [comment]
Seems like we should update the `@horner` in Base with the new one. Thoughts from others?

**Comment by simonbyrne** [comment]
> Seems like we should update the `@horner` in Base with the new one. Thoughts from others?

Well, technically it would no longer be a Horner scheme (there is a chance that accuracy guarantees are predicated on it being evaluated in this manner)

**Comment by musm** [comment]
I think Viral meant update `@evalpoly`... 
Sometime soon I'm going to update  the Estrin scheme code (it's not efficient yet) and compare ( but I suspect the simple split scheme may still perform best)

**Comment by simonbyrne** [comment]
For help with these sort of low-level optimisations, you might want to check out https://github.com/carnaval/IACA.jl

**Comment by musm** [comment]
Edit: sensible benchmark results posted above. Looks like the overhead of setting up SIMD doesn't pay off until the polynomial is  order 18.

**Comment by eschnett** [comment]
Did you try setting up more than 2 independent instruction streams? E.g. using 4 streams might be even faster, given that multiplication and addition has several cycles of latency.

**Comment by augustt198** [comment]
Reviving this – I've started a (WIP) package for SIMD polynomial evaluation here: https://github.com/augustt198/SIMDPoly.jl. Using 4-wide vectors with real inputs it seems SIMD becomes advantageous around degree 12, if my benchmarking is correct. SIMD polynomial division with complex inputs are around 2x faster than `evalpoly(z, num) / evalpoly(z, denom)`.

SIMD.jl is a dependency so I don't think there's much that could be transferred to Base.

**Comment by musm** [comment]
@augustt198  nice work

In your benchmark code https://github.com/augustt198/SIMDPoly.jl/blob/master/test/benchmark2.jl realistically it should also  be timing the time it takes to pack the coefficients as well if it's to compare fairly.

The timings  including packing   highlight what @eschnett  mentioned back in https://github.com/JuliaLang/julia/issues/18410#issuecomment-245922269   they become hopelessly slow for the SIMD version 😢

**Comment by augustt198** [comment]
Sure I can add a disclaimer about that. I'm mostly interested in cases where the coefficients are constants: Taylor series, rational approximations, etc.

**Comment by musm** [comment]
True, I guess in practice that's not the main bottleneck. So perhaps just ignore my comment then 😄
